### PR TITLE
Pippa/log attn mask

### PR DESCRIPTION
**Pull request by pippaduckett1**
Log attention mask image to comet

### CODE REVIEW

File: src/hepattn/models/wrapper.py [hunk 1]
```diff
@@ -177,3 +178,27 @@ def mlt_opt(self, losses, outputs):
             mtl_backward(losses=layer_losses, features=layer_features, aggregator=UPGrad())
 
         opt.step()
+
+    def log_figure(self, name, fig, step=None):
```
Comment (samvanstroud) [inline src/hepattn/models/wrapper.py:182]: As far as I know everyone is using comet, can we remove some of the structure here and just call `log_image` directly from the callback? I don't think `wrapper.py` should import `matplotlib`, but it's fine if the callback does

File: src/hepattn/models/maskformer.py [hunk 2]
```diff
@@ -161,6 +204,16 @@ def forward(self, inputs: dict[str, Tensor]) -> dict[str, Tensor]:
             else:
                 attn_mask = None
 
+            if (
+                self.log_attn_mask
+                and (attn_mask is not None)
+                and (self.step_ % 1000 == 0)
+            ):
```
[the remark this answers is not in the document]
Comment (samvanstroud) [reply src/hepattn/models/maskformer.py:195]: I think we need some way of storing the attention mask so it can be accessed in the callback, unless you had another suggestion @trahxam? Returning it would be a bit more involved...

File: src/hepattn/models/maskformer.py [hunk 3]
```diff
@@ -61,10 +61,37 @@ def __init__(
...
+        return None
+
+    def clear_last_attention_mask(self):
+        """Clear the stored attention mask after logging."""
+        if hasattr(self, '_last_attn_mask'):
+            del self._last_attn_mask
+        if hasattr(self, '_last_attn_mask_step'):
+            del self._last_attn_mask_step
+        if hasattr(self, '_last_attn_mask_layer'):
+            del self._last_attn_mask_layer
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:89]: I would remove the boilerplate here and just directly access/clear the member from the callback.

File: src/hepattn/callbacks/attn_mask_logger.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        
+        if step % 1000 == 0:
+            attn_mask_info = model.get_last_attention_mask() if hasattr(model, "get_last_attention_mask") else None
+            
+            if attn_mask_info is not None:
+                attn_mask, step, layer = attn_mask_info
+                plt.figure(constrained_layout=True, dpi=300)
+                plt.imshow(attn_mask.numpy(), aspect="auto")
+                pl_module.log_figure(f"local_ca_mask_step{step}_layer{layer}", plt.gcf(), step=step)
+                plt.close()
```
Comment (samvanstroud) [inline src/hepattn/callbacks/attn_mask_logger.py:30]: Can we factor out the actual plotting code to a function that is called by both the train and val hooks?

File: src/hepattn/callbacks/attn_mask_logger.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+class AttnMaskLogger(Callback):
+    def on_validation_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx=0):
+        model = pl_module.model if hasattr(pl_module, "model") else pl_module
+        attn_mask_info = model.get_last_attention_mask() if hasattr(model, "get_last_attention_mask") else None
+        
+        if attn_mask_info is not None:
+            attn_mask, step, layer = attn_mask_info
+            plt.figure(constrained_layout=True, dpi=300)
+            plt.imshow(attn_mask.numpy(), aspect="auto")
```
Comment (samvanstroud) [inline src/hepattn/callbacks/attn_mask_logger.py:12]: can you add a legend or colourbar to each image which indicates what `0` and `1` represent in terms of "used in attention", "masked", just so there is zero ambiguity

**Review by samvanstroud — COMMENTED** [review]
Thanks @pippaduckett1, this looks good! I left a few comments. I think we can avoid the changes to `wrapper.py`, and the linting needs a look.

File: src/hepattn/models/maskformer.py [hunk 6]
```diff
@@ -161,6 +164,16 @@ def forward(self, inputs: dict[str, Tensor]) -> dict[str, Tensor]:
...
 
+            if (
+                self.log_attn_mask
+                and (attn_mask is not None)
+                and (self.step_ % 1000 == 0)
+            ):
+                # Store for callback to log later
+                self._last_attn_mask = attn_mask[0].detach().cpu().clone()
+                self._last_attn_mask_step = self.step_
+                self._last_attn_mask_layer = layer_index
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:175]: it looks like we will always log the final layer mask at the moment? is it worth logging e.g. the first and last?

File: src/hepattn/models/maskformer.py [hunk 7]
```diff
@@ -61,10 +61,13 @@ def __init__(
         self.input_sort_field = input_sort_field
         self.use_attn_masks = use_attn_masks
         self.use_query_masks = use_query_masks
+        self.log_attn_mask = log_attn_mask
+        self.step_ = 0
```
Comment (samvanstroud) [inline src/hepattn/models/maskformer.py:65]: btw I think normally leading underscores are used to indicate internal use, not trailing.

**Review by samvanstroud — COMMENTED** [review]
Thanks for the updates, spotted a few more small things
